trace/sqs/message.go

```diff
@@ -0,0 +1,105 @@
+package sqs
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/sqs"
+)
+
+const attributeDataTypeString string = "String"
+
+// MessageBuilder helps building messages to be sent to SQS.
+type MessageBuilder struct {
+	err   error
+	input *sqs.SendMessageInput
+}
+
+// NewMessageBuilder creates a new MessageBuilder that helps creating messages.
+func NewMessageBuilder() *MessageBuilder {
+	return &MessageBuilder{
+		input: &sqs.SendMessageInput{
+			MessageAttributes: map[string]*sqs.MessageAttributeValue{},
+		},
+	}
+}
+
+// Message is a struct embedding information about messages that will
+// be later published to SQS thanks to the SQS publisher.
+type Message struct {
+	input *sqs.SendMessageInput
+}
+
+// Build tries to build a message given its specified data and returns an error if any goes wrong.
+func (b *MessageBuilder) Build() (*Message, error) {
+	if b.err != nil {
+		return nil, b.err
+	}
+
+	if b.input.MessageBody == nil {
+		return nil, errors.New("missing required field: message body")
+	}
+
+	if b.input.QueueUrl == nil {
+		return nil, errors.New("missing required field: message queue URL")
+	}
+
+	// Messages with either a group ID or deduplication ID can't have a delay.
+	// These two attributes are only used for FIFO queues, which don't allow for individual message delays.
+	if (b.input.MessageGroupId != nil || b.input.MessageDeduplicationId != nil) && b.input.DelaySeconds != nil {
```

Comment:
Is this not checked by the package? How does the package behave when these are set?

Reply:
Unfortunately, it does not check it and returns an empty msg ID. So checking it here avoids sending an unnecessary request.